File: grading.py
import json


def normalize(s: str) -> str:
    return " ".join((s or "").strip().lower().split())


def grade_answer(question: dict, raw_answer) -> tuple[bool | None, float, bool]:
    """
    Returns (is_correct, points_earned, needs_manual).
    is_correct may be None for ungradable (e.g., poll, open-ended).
    """
    qtype = question["type"]
    points = float(question.get("points") or 1)
    correct = question.get("correct_answers")
    if isinstance(correct, str):
        try:
            correct = json.loads(correct)
        except Exception:
            correct = []
    correct = correct or []

    if raw_answer is None or raw_answer == "" or raw_answer == []:
        return (False if qtype in _AUTO_TYPES else None, 0.0, qtype in _MANUAL_TYPES)

    if qtype == "mcq_single":
        try:
            picked = int(raw_answer) if not isinstance(raw_answer, list) else int(raw_answer[0])
        except Exception:
            return False, 0.0, False
        ok = picked in correct
        return ok, points if ok else 0.0, False

    if qtype == "mcq_multi":
        picked = raw_answer if isinstance(raw_answer, list) else [raw_answer]
        try:
            picked = sorted({int(x) for x in picked})
        except Exception:
            return False, 0.0, False
        ok = picked == sorted(set(int(x) for x in correct))
        return ok, points if ok else 0.0, False

    if qtype == "true_false":
        try:
            picked = int(raw_answer) if not isinstance(raw_answer, list) else int(raw_answer[0])
        except Exception:
            return False, 0.0, False
        ok = picked in correct
        return ok, points if ok else 0.0, False

    if qtype in ("short_answer", "fill_blank"):
        val = raw_answer if isinstance(raw_answer, str) else (raw_answer[0] if raw_answer else "")
        accepted = [normalize(c) for c in correct]
        ok = normalize(val) in accepted if accepted else None
        if ok is None:
            return None, 0.0, True
        return ok, points if ok else 0.0, False

    if qtype == "long_answer":
        return None, 0.0, True

    if qtype == "matching":
        # raw_answer: list of strings (the picked "b" for each prompt index)
        # correct = [<right side of each pair>] in same order as options
        # options = [{"a": "...", "b": "..."}, ...]
        options = question.get("options")
        if isinstance(options, str):
            try:
                options = json.loads(options)
            except Exception:
                options = []
        options = options or []
        if not isinstance(raw_answer, list):
            return False, 0.0, False
        ok_count = 0
        for i, pair in enumerate(options):
            expected = (pair or {}).get("b", "") if isinstance(pair, dict) else ""
            picked = raw_answer[i] if i < len(raw_answer) else ""
            if normalize(picked) == normalize(expected):
                ok_count += 1
        # Partial credit: proportion of pairs matched
        if not options:
            return False, 0.0, False
        ratio = ok_count / len(options)
        return ratio == 1.0, points * ratio, False

    if qtype == "ordering":
        # raw_answer: list of indices = student's ordering of the options
        # correct: list of indices = correct ordering
        options = question.get("options")
        if isinstance(options, str):
            try: options = json.loads(options)
            except Exception: options = []
        options = options or []
        if not isinstance(raw_answer, list):
            return False, 0.0, False
        correct_order = correct if correct else list(range(len(options)))
        try:
            picked = [int(x) for x in raw_answer]
        except Exception:
            return False, 0.0, False
        ok = picked == correct_order
        return ok, points if ok else 0.0, False

    if qtype == "drag_drop":
        # options = [{"item":"...","bin":"..."}, ...]
        # raw_answer = list of strings (picked bin for each item)
        options = question.get("options")
        if isinstance(options, str):
            try: options = json.loads(options)
            except Exception: options = []
        options = options or []
        if not isinstance(raw_answer, list):
            return False, 0.0, False
        ok_count = 0
        for i, pair in enumerate(options):
            expected = (pair or {}).get("bin", "") if isinstance(pair, dict) else ""
            picked = raw_answer[i] if i < len(raw_answer) else ""
            if normalize(picked) == normalize(expected):
                ok_count += 1
        if not options:
            return False, 0.0, False
        ratio = ok_count / len(options)
        return ratio == 1.0, points * ratio, False

    if qtype == "hotspot":
        # options = [{"x": float 0..1, "y": float 0..1, "r": float 0..1, "label": str}]
        # raw_answer: {"x": float, "y": float}
        options = question.get("options")
        if isinstance(options, str):
            try: options = json.loads(options)
            except Exception: options = []
        options = options or []
        if not isinstance(raw_answer, dict):
            return False, 0.0, False
        try:
            x = float(raw_answer.get("x", -1))
            y = float(raw_answer.get("y", -1))
        except Exception:
            return False, 0.0, False
        for hs in options:
            try:
                hx, hy, hr = float(hs.get("x")), float(hs.get("y")), float(hs.get("r", 0.05))
            except Exception:
                continue
            if ((x - hx) ** 2 + (y - hy) ** 2) ** 0.5 <= hr:
                return True, points, False
        return False, 0.0, False

    # poll, rating, open_ended, word_cloud — not graded
    return None, 0.0, False


_AUTO_TYPES = {"mcq_single", "mcq_multi", "true_false", "short_answer", "fill_blank"}
_MANUAL_TYPES = {"long_answer"}


QUESTION_TYPES = [
    ("mcq_single", "Multiple Choice (single answer)"),
    ("mcq_multi", "Multiple Choice (multiple answers)"),
    ("true_false", "True / False (or Yes / No)"),
    ("short_answer", "Short Answer"),
    ("long_answer", "Long Answer / Essay"),
    ("fill_blank", "Fill in the Blank"),
    ("rating", "Rating (1–5 stars)"),
    ("nps", "NPS — Net Promoter Score (0–10)"),
    ("poll", "Poll choice (no correct answer)"),
    ("open_ended", "Open-ended (free text, ungraded)"),
    ("word_cloud", "Word Cloud (live)"),
    # Form fields (for kind='form' — JotForm-style data collection)
    ("email", "Email address (form field)"),
    ("phone", "Phone number (form field)"),
    ("date", "Date picker (form field)"),
    ("number", "Number (form field)"),
    ("dropdown", "Dropdown / Select (form field)"),
    # Complex interactive types
    ("matching", "Matching (pair left side with right side)"),
    ("ordering", "Ordering / Sequencing (arrange in correct order)"),
    ("drag_drop", "Drag & Drop (sort items into categories)"),
    ("hotspot", "Image Hotspot (click on the correct area)"),
]

File: php/includes/grading.php
<?php
/**
 * Grading engine + question-type catalogue. Ported from the Python grading.py.
 *
 * grade_answer($question, $answer) returns [is_correct, points_earned, needs_manual]
 * where is_correct may be null for ungradable types (polls, open text).
 */

declare(strict_types=1);

/** All supported types: [value, label, group]. Group drives editor sections. */
function question_types(): array
{
    return [
        // group: choice
        ['mcq_single', 'Multiple Choice (single answer)', 'choice'],
        ['mcq_multi', 'Multiple Choice (multiple answers)', 'choice'],
        ['true_false', 'True / False (or Yes / No)', 'choice'],
        ['dropdown', 'Dropdown / Select', 'choice'],
        // group: text
        ['short_answer', 'Short Answer', 'text'],
        ['long_answer', 'Long Answer / Essay', 'text'],
        ['fill_blank', 'Fill in the Blank', 'text'],
        // group: scale
        ['rating', 'Rating (1–5 stars)', 'scale'],
        ['nps', 'NPS — Net Promoter Score (0–10)', 'scale'],
        ['slider', 'Slider / Range', 'scale'],
        // group: poll
        ['poll', 'Poll choice (no correct answer)', 'poll'],
        ['open_ended', 'Open-ended (free text, ungraded)', 'poll'],
        ['word_cloud', 'Word Cloud (live)', 'poll'],
        // group: interactive
        ['matching', 'Matching (pair left with right)', 'interactive'],
        ['ordering', 'Ordering / Sequencing', 'interactive'],
        ['drag_drop', 'Drag & Drop (sort into categories)', 'interactive'],
        ['hotspot', 'Image Hotspot (click the area)', 'interactive'],
        // group: form fields
        ['email', 'Email address', 'form'],
        ['phone', 'Phone number', 'form'],
        ['number', 'Number', 'form'],
        ['date', 'Date picker', 'form'],
        ['time', 'Time picker', 'form'],
        ['datetime', 'Date & Time', 'form'],
        ['url', 'URL / Website', 'form'],
        ['address', 'Full address', 'form'],
        ['full_name', 'Full name (first + last)', 'form'],
        ['file_upload', 'File upload', 'form'],
        ['signature', 'Signature pad', 'form'],
        ['consent', 'Terms / Consent checkbox', 'form'],
        ['section_break', 'Section break / instructions', 'form'],
    ];
}

/** Types that are auto-graded to correct/incorrect. */
function auto_graded_types(): array
{
    return ['mcq_single', 'mcq_multi', 'true_false', 'short_answer', 'fill_blank', 'dropdown'];
}

/** Types that always need manual grading. */
function manual_types(): array
{
    return ['long_answer'];
}

/** Types that have an options list (choices). */
function choice_types(): array
{
    return ['mcq_single', 'mcq_multi', 'true_false', 'dropdown', 'poll'];
}

function normalize_text(?string $s): string
{
    return implode(' ', preg_split('/\s+/', trim(mb_strtolower((string)$s))));
}

/**
 * Grade a single answer.
 * @param array $question  row with type/options/correct_answers/points
 * @param mixed $answer    decoded submitted value
 * @return array [?bool is_correct, float points_earned, bool needs_manual]
 */
function grade_answer(array $question, $answer): array
{
    $type = $question['type'];
    $points = (float)($question['points'] ?? 1);
    $correct = $question['correct_answers'] ?? [];
    if (is_string($correct)) $correct = json_col($correct, []);
    $correct = is_array($correct) ? $correct : [];
    $options = $question['options'] ?? [];
    if (is_string($options)) $options = json_col($options, []);
    $options = is_array($options) ? $options : [];

    $empty = ($answer === null || $answer === '' || $answer === []);
    if ($empty) {
        $isAuto = in_array($type, auto_graded_types(), true);
        return [$isAuto ? false : null, 0.0, in_array($type, manual_types(), true)];
    }

    switch ($type) {
        case 'mcq_single':
        case 'true_false':
        case 'dropdown':
            $picked = is_array($answer) ? ($answer[0] ?? null) : $answer;
            $pi = to_int($picked, null);
            if ($pi === null) return [false, 0.0, false];
            $ok = in_array($pi, int_list($correct), true);
            return [$ok, $ok ? $points : 0.0, false];

        case 'mcq_multi':
            $picked = is_array($answer) ? $answer : [$answer];
            $pi = int_list($picked); sort($pi);
            $ci = int_list($correct); sort($ci);
            $ok = ($pi === $ci);
            return [$ok, $ok ? $points : 0.0, false];

        case 'short_answer':
        case 'fill_blank':
            $val = is_array($answer) ? ($answer[0] ?? '') : $answer;
            $accepted = array_map('normalize_text', $correct);
            if (!$accepted) return [null, 0.0, true];
            $ok = in_array(normalize_text((string)$val), $accepted, true);
            return [$ok, $ok ? $points : 0.0, false];

        case 'long_answer':
            return [null, 0.0, true];

        case 'matching':
            // options: [{a,b}], answer: [picked b for each index]
            if (!is_array($answer) || !$options) return [false, 0.0, false];
            $okc = 0;
            foreach ($options as $i => $pair) {
                $exp = is_array($pair) ? ($pair['b'] ?? '') : '';
                $pick = $answer[$i] ?? '';
                if (normalize_text((string)$pick) === normalize_text((string)$exp)) $okc++;
            }
            $ratio = count($options) ? $okc / count($options) : 0;
            return [$ratio == 1.0, $points * $ratio, false];

        case 'ordering':
            if (!is_array($answer) || !$options) return [false, 0.0, false];
            $order = $correct ? int_list($correct) : range(0, count($options) - 1);
            $pi = int_list($answer);
            $ok = ($pi === $order);
            return [$ok, $ok ? $points : 0.0, false];

        case 'drag_drop':
            // options: [{item,bin}], answer: [picked bin per item]
            if (!is_array($answer) || !$options) return [false, 0.0, false];
            $okc = 0;
            foreach ($options as $i => $pair) {
                $exp = is_array($pair) ? ($pair['bin'] ?? '') : '';
                $pick = $answer[$i] ?? '';
                if (normalize_text((string)$pick) === normalize_text((string)$exp)) $okc++;
            }
            $ratio = count($options) ? $okc / count($options) : 0;
            return [$ratio == 1.0, $points * $ratio, false];

        case 'hotspot':
            // options: [{x,y,r}], answer: {x,y}
            if (!is_array($answer)) return [false, 0.0, false];
            $x = (float)($answer['x'] ?? -1);
            $y = (float)($answer['y'] ?? -1);
            foreach ($options as $hs) {
                $hx = (float)($hs['x'] ?? 0); $hy = (float)($hs['y'] ?? 0); $hr = (float)($hs['r'] ?? 0.05);
                if (sqrt(($x-$hx)**2 + ($y-$hy)**2) <= $hr) return [true, $points, false];
            }
            return [false, 0.0, false];

        default:
            // poll, rating, nps, form fields — collected, not graded
            return [null, 0.0, false];
    }
}
